### ISSUE TITLE
Make error in Ubuntu 16.04

### ISSUE THREAD

**Issue by YongTingDDak** [facebook/folly#546]
I Tried to make folly, but it doesnt work this message

```
libtool: compile:  g++ -DHAVE_CONFIG_H -I./.. -pthread -I/usr/include -I/usr/local/include -I/usr/local/boost_1_63_0 -std=gnu++1y -g -O2 -MT io/async/AsyncSSLSocket.lo -MD -MP -MF io/async/.deps/AsyncSSLSocket.Tpo -c io/async/AsyncSSLSocket.cpp  -fPIC -DPIC -o io/async/.libs/AsyncSSLSocket.o
## # **io/async/AsyncSSLSocket.cpp:180:12: error: aggregate ‘BIO_METHOD {anonymous}::sslBioMethod’ has incomplete type and cannot be defined
 BIO_METHOD sslBioMethod; // error
            ^**
io/async/AsyncSSLSocket.cpp: In member function ‘void folly::AsyncSSLSocket::attachSSLContext(const std::shared_ptr<folly::SSLContext>&)’:
io/async/AsyncSSLSocket.cpp:484:7: error: invalid use of incomplete type ‘SSL {aka struct ssl_st}’
   ssl_->initial_ctx = sslCtx;
       ^
In file included from /usr/local/include/openssl/crypto.h:31:0,
                 from /usr/local/include/openssl/bio.h:20,
                 from /usr/local/include/openssl/asn1.h:16,
                 from ./../folly/io/async/ssl/OpenSSLPtrTypes.h:21,
                 from ./../folly/io/async/AsyncTransport.h:25,
                 from ./../folly/io/async/AsyncSocket.h:26,
                 from ./../folly/io/async/AsyncSSLSocket.h:23,
                 from io/async/AsyncSSLSocket.cpp:17:
/usr/local/include/openssl/ossl_typ.h:144:16: note: forward declaration of ‘SSL {aka struct ssl_st}’
 typedef struct ssl_st SSL;
                ^
io/async/AsyncSSLSocket.cpp: In member function ‘void folly::AsyncSSLSocket::detachSSLContext()’:
io/async/AsyncSSLSocket.cpp:503:11: error: invalid use of incomplete type ‘SSL {aka struct ssl_st}’
   if (ssl_->initial_ctx) {
           ^
In file included from /usr/local/include/openssl/crypto.h:31:0,
                 from /usr/local/include/openssl/bio.h:20,
                 from /usr/local/include/openssl/asn1.h:16,
                 from ./../folly/io/async/ssl/OpenSSLPtrTypes.h:21,
                 from ./../folly/io/async/AsyncTransport.h:25,
                 from ./../folly/io/async/AsyncSocket.h:26,
                 from ./../folly/io/async/AsyncSSLSocket.h:23,
                 from io/async/AsyncSSLSocket.cpp:17:
/usr/local/include/openssl/ossl_typ.h:144:16: note: forward declaration of ‘SSL {aka struct ssl_st}’
 typedef struct ssl_st SSL;
                ^
io/async/AsyncSSLSocket.cpp:504:22: error: invalid use of incomplete type ‘SSL {aka struct ssl_st}’
     SSL_CTX_free(ssl_->initial_ctx);
                      ^
In file included from /usr/local/include/openssl/crypto.h:31:0,
                 from /usr/local/include/openssl/bio.h:20,
                 from /usr/local/include/openssl/asn1.h:16,
                 from ./../folly/io/async/ssl/OpenSSLPtrTypes.h:21,
                 from ./../folly/io/async/AsyncTransport.h:25,
                 from ./../folly/io/async/AsyncSocket.h:26,
                 from ./../folly/io/async/AsyncSSLSocket.h:23,
                 from io/async/AsyncSSLSocket.cpp:17:
/usr/local/include/openssl/ossl_typ.h:144:16: note: forward declaration of ‘SSL {aka struct ssl_st}’
 typedef struct ssl_st SSL;
                ^
io/async/AsyncSSLSocket.cpp:505:9: error: invalid use of incomplete type ‘SSL {aka struct ssl_st}’
     ssl_->initial_ctx = nullptr;
         ^
In file included from /usr/local/include/openssl/crypto.h:31:0,
                 from /usr/local/include/openssl/bio.h:20,
                 from /usr/local/include/openssl/asn1.h:16,
                 from ./../folly/io/async/ssl/OpenSSLPtrTypes.h:21,
                 from ./../folly/io/async/AsyncTransport.h:25,
                 from ./../folly/io/async/AsyncSocket.h:26,
                 from ./../folly/io/async/AsyncSSLSocket.h:23,
                 from io/async/AsyncSSLSocket.cpp:17:
/usr/local/include/openssl/ossl_typ.h:144:16: note: forward declaration of ‘SSL {aka struct ssl_st}’
 typedef struct ssl_st SSL;
                ^
io/async/AsyncSSLSocket.cpp: In member function ‘bool folly::AsyncSSLSocket::isServerNameMatch() const’:
io/async/AsyncSSLSocket.cpp:553:9: error: invalid use of incomplete type ‘SSL_SESSION {aka struct ssl_session_st}’
   if(!ss->tlsext_hostname) {
         ^
In file included from ./../folly/io/async/ssl/OpenSSLPtrTypes.h:30:0,
                 from ./../folly/io/async/AsyncTransport.h:25,
                 from ./../folly/io/async/AsyncSocket.h:26,
                 from ./../folly/io/async/AsyncSSLSocket.h:23,
                 from io/async/AsyncSSLSocket.cpp:17:
/usr/local/include/openssl/ssl.h:228:16: note: forward declaration of ‘SSL_SESSION {aka struct ssl_session_st}’
 typedef struct ssl_session_st SSL_SESSION;
                ^
io/async/AsyncSSLSocket.cpp:556:37: error: invalid use of incomplete type ‘SSL_SESSION {aka struct ssl_session_st}’
   return (tlsextHostname_.compare(ss->tlsext_hostname) ? false : true);
                                     ^
In file included from ./../folly/io/async/ssl/OpenSSLPtrTypes.h:30:0,
                 from ./../folly/io/async/AsyncTransport.h:25,
                 from ./../folly/io/async/AsyncSocket.h:26,
                 from ./../folly/io/async/AsyncSSLSocket.h:23,
                 from io/async/AsyncSSLSocket.cpp:17:
/usr/local/include/openssl/ssl.h:228:16: note: forward declaration of ‘SSL_SESSION {aka struct ssl_session_st}’
 typedef struct ssl_session_st SSL_SESSION;
                ^
io/async/AsyncSSLSocket.cpp: In member function ‘const char* folly::AsyncSSLSocket::getSSLCertSigAlgName() const’:
io/async/AsyncSSLSocket.cpp:877:31: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
     int nid = OBJ_obj2nid(cert->sig_alg->algorithm);
                               ^
In file included from /usr/local/include/openssl/crypto.h:31:0,
                 from /usr/local/include/openssl/bio.h:20,
                 from /usr/local/include/openssl/asn1.h:16,
                 from ./../folly/io/async/ssl/OpenSSLPtrTypes.h:21,
                 from ./../folly/io/async/AsyncTransport.h:25,
                 from ./../folly/io/async/AsyncSocket.h:26,
                 from ./../folly/io/async/AsyncSSLSocket.h:23,
                 from io/async/AsyncSSLSocket.cpp:17:
/usr/local/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^
Makefile:1703: recipe for target 'io/async/AsyncSSLSocket.lo' failed
make[1]: *** [io/async/AsyncSSLSocket.lo] Error 1
make[1]: Leaving directory '/home/lab-com/folly-master/folly'
Makefile:1795: recipe for target 'check-recursive' failed
make: *** [check-recursive] Error 1

========================================================================
```
it has error in AsyncSSLSocket.cpp i cannot solve this problem. help plz

**Comment by KamalGalrani** [comment]
How did you solve the issue? I am facing the same error

**Comment by fredemmott** [comment]
@KamalGalrani : on 16.04 you can install the gcc-5 packages instead

**Comment by fredemmott** [comment]
might also be openssl v1.0 vs 1.1

**Comment by fredemmott** [comment]
On a clean debian 9 image, I get:

```
/tmp/out/hhvm-3.20.2/third-party/folly/src/folly/io/async/AsyncSSLSocket.cpp: In member function 'bool folly::AsyncSSLSocket::isServerNameMatch() const':
/tmp/out/hhvm-3.20.2/third-party/folly/src/folly/io/async/AsyncSSLSocket.cpp:552:9: error: invalid use of incomplete type 'SSL_SESSION {aka struct ssl_session_st}'
   if(!ss->tlsext_hostname) {
         ^~
In file included from /tmp/out/hhvm-3.20.2/third-party/folly/src/folly/io/async/ssl/OpenSSLPtrTypes.h:30:0,
                 from /tmp/out/hhvm-3.20.2/third-party/folly/src/folly/io/async/AsyncTransport.h:25,
                 from /tmp/out/hhvm-3.20.2/third-party/folly/src/folly/io/async/AsyncSocket.h:26,
                 from /tmp/out/hhvm-3.20.2/third-party/folly/src/folly/io/async/AsyncSSLSocket.h:23,
                 from /tmp/out/hhvm-3.20.2/third-party/folly/src/folly/io/async/AsyncSSLSocket.cpp:17:
/usr/include/openssl/ssl.h:228:16: note: forward declaration of 'SSL_SESSION {aka struct ssl_session_st}'
 typedef struct ssl_session_st SSL_SESSION;
                ^~~~~~~~~~~~~~
/tmp/out/hhvm-3.20.2/third-party/folly/src/folly/io/async/AsyncSSLSocket.cpp:555:37: error: invalid use of incomplete type 'SSL_SESSION {aka struct ssl_session_st}'
   return (tlsextHostname_.compare(ss->tlsext_hostname) ? false : true);
                                     ^~
In file included from /tmp/out/hhvm-3.20.2/third-party/folly/src/folly/io
```

**Comment by yfeldblum** [comment]
Closing because this issue is too old.
